cost_rules/app.py

```diff
@@ -0,0 +1,293 @@
+                'DimensionName': 'TAG',
+                'DimensionKey': t,
+            }
+        }
+        rules.append(rule)
+
+    return rules
+
+def rule_generator(params):
+    '''
+        Input:
+
+        InheritedValues:
+          TagOrder:
+            - CostCenterOther
+            - CostCenter
+          RulePosition: Last
+        RegularValues:
+          - Value: Category One
+            Accounts:
+              - 123abc
+              - 456xyz
+            TagNames:
+              - CostCenter
+              - CostCenterOther
+            TagEndsWith:
+              - 123400
+          - Value: Category Two
+            ...
+    '''
+    # Rule order matters, first match wins and rules should
+    # generally be ordered most-common to least-common
+    rules = []
+
+    # First get the inherited tag info
+    it_position = 'Last'  # either 'First' or 'Last'
+    it_rules = []
+    if 'InheritedValues' in params:
+        info = params['InheritedValues']
+
+        if 'RulePosition' in info:
+            it_position = info['RulePosition']
+
+        if 'TagOrder' in info:
+            order = info['TagOrder']
+            it_rules = inherited_rules(order)
+
+    # add our inherited tag rules if their position is First
+    if it_rules and it_position == 'First':
+        rules.extend(it_rules)
+
+    # iterate over our labels, building the rules for each
+    if 'RegularValues' in params:
+        for info in params['RegularValues']:
+
+            # First add tag suffix rules
+            tags = None
+            if 'TagNames' in info:
+                tags = info['TagNames']
+                if 'TagEndsWith' in info:
+                    rules.extend(tag_rules(info['Value'], tags, info['TagEndsWith'], 'ENDS_WITH'))
+                if 'TagStartsWith' in info:
+                    rules.extend(tag_rules(info['Value'], tags, info['TagStartsWith'], 'STARTS_WITH'))
+
+            # Then add an account rule
+            if 'Accounts' in info:
+                rules.append(account_rule(info['Value'], info['Accounts'], tags))
+
+    # add our inherited tag rules if their position is Last
+    if it_rules and it_position == 'Last':
+        rules.extend(it_rules)
+
+    return rules
+
+def handler(event, context):
+    """Sample pure Lambda function
+
+    Parameters
+    ----------
+    event: dict, required
+        Macro Input Format
+        https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html
+
+    context: object, required
+        Lambda Context runtime methods and attributes
+
+        Context doc: https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html
+
+    Returns
+    ------
+        Macro Output Format
+        https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html
+    """
+
+    out = {
+        'requestId': None,
+        'status': None,
+        'fragment': None
+    }
+
+    try:
+        params = event['fragment']
```

Comment:
you should be able to get just the params from the event with `event['params']`

Reply:
What we want to use as input is the macro fragment because the allowed types for parameters don't meet our needs.